compiler/noirc_evaluator/src/ssa/opt/inline_const_brillig_calls.rs

```diff
@@ -0,0 +1,463 @@
+use std::collections::{BTreeMap, HashSet};
+
+use acvm::{
+    blackbox_solver::StubbedBlackBoxSolver,
+    brillig_vm::{MemoryValue, VMStatus, VM},
+    FieldElement,
+};
+
+use im::Vector;
+
+use crate::{
+    brillig::{
+        brillig_gen::{brillig_fn::FunctionContext, gen_brillig_for},
+        Brillig,
+    },
+    ssa::{
+        ir::{
+            basic_block::BasicBlockId,
+            dfg::DataFlowGraph,
+            function::{Function, FunctionId, RuntimeType},
+            instruction::{Instruction, InstructionId},
+            types::Type,
+            value::{Value, ValueId},
+        },
+        Ssa,
+    },
+};
+
+impl Ssa {
+    pub(crate) fn inline_const_brillig_calls(mut self, brillig: &Brillig) -> Self {
+        // Collect all brillig functions so that later we can find them when processing a call instruction
+        let mut brillig_functions: BTreeMap<FunctionId, Function> = BTreeMap::new();
+        for (func_id, func) in &self.functions {
+            if let RuntimeType::Brillig(..) = func.runtime() {
+                let cloned_function = Function::clone_with_id(*func_id, func);
+                brillig_functions.insert(*func_id, cloned_function);
+            };
+        }
+
+        // Keep track of which brillig functions we couldn't completely inline: we'll remove the ones we could.
+        let mut brillig_functions_we_could_not_inline = HashSet::new();
+
+        for func in self.functions.values_mut() {
+            func.inline_const_brillig_calls(
+                brillig,
+                &brillig_functions,
+                &mut brillig_functions_we_could_not_inline,
+            );
+        }
+
+        // Remove the brillig functions that are no longer called
+        for func_id in brillig_functions.keys() {
+            // We never want to remove the main function (it could be `unconstrained` or it
+            // could have been turned into brillig if `--force-brillig` was given)
+            if self.main_id == *func_id {
+                continue;
+            }
+
+            if brillig_functions_we_could_not_inline.contains(func_id) {
+                continue;
+            }
+
+            // We also don't want to remove entry points
+            if self.entry_point_to_generated_index.contains_key(func_id) {
+                continue;
+            }
+
+            self.functions.remove(func_id);
+        }
+
+        self
+    }
+}
+
+/// Result of trying to evaluate an instruction (any instruction) in this pass.
+enum EvaluationResult {
+    /// Nothing was done because the instruction wasn't a call to a brillig function,
+    /// or some arguments to it were not constants.
+    NotABrilligCall,
+    /// The instruction was a call to a brillig function, but we couldn't evaluate it.
```

Comment:
```suggestion
    /// The instruction was a call to a brillig function, but we couldn't evaluate it.
    /// This can occur in the situation where the brillig function reaches a "trap" or a foreign call opcode.
```